specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,141 @@
+- `input_len` Length for correct padding
+- `acc_len` How many bytes we have processed
+- `condition_80_inv` The inverse of `input_len - acc_len` or 0.
+- `padded_byte` Mostly the same as the original `byte` but padded
+- `is_pad_zone` A flag to define the rows that `byte` should be 0
+
+| offset | byte | input_len | acc_len | condition_80_inv | padded_byte | is_pad_zone |
+| -----: | :--- | --------: | ------: | ---------------: | :---------- | ----------: |
+|      0 | 0xff |       250 |     136 |                  | 0xff        |           0 |
+|      1 | 0xff |       250 |     137 |                  | 0xff        |           0 |
+|      2 | 0xff |       250 |     138 |                  | 0xff        |           0 |
+|      3 | 0xff |       ... |     ... |                  | 0xff        |           0 |
+|      4 | 0xff |       250 |     249 |                1 | 0xff        |           0 |
+|      5 | 0xff |       250 |     250 |                0 | 0xff        |           0 |
+|      6 | 0x00 |       250 |     251 |               -1 | 0x80        |           1 |
+|    ... | ...  |       ... |     ... |              ... | ...         |         ... |
+|    134 | 0x00 |       250 |     270 |                  | 0x00        |           1 |
+|    135 | 0x00 |       250 |     271 |                  | 0x01        |           1 |
+
+#### Checks
+
+Generally we want these properties:
+
+1. `byte` after `input_len` should be 0, so that prover doesn't cheat about the input length. For exapmle, the prover specify an input length of 100 bytes, so that it shouldn't have a non-zero byte at position 120.
+2. The padding `0x80`, `0x01`, or `0x81` are placed at the correct place.
+
+We apply two different checks on the 0~134-th rows and the 135th row.
+
+1. For 0-th row
+   1. `input_len` is copied from the Lookup Region
+   2. `acc_len` is copied from the Lookup Region
+   3. `curr.is_pad_zone === 0`
+2. For all rows
+   1. If `is_pad_zone` then `byte === 0`. `is_pad_zone * byte === 0`
+3. For 0~134-th rows
+   1. `next.input_len === curr.input_len`
+   2. `next.acc_len === curr.acc_len + 1`
+   3. Inverse check for `curr.condition_80_inv`
+   4. If `curr.input_len - curr.acc_len` is 0, pad `0x80`: `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80`
+4. For 1~135th rows
+   1. Set `is_pad_zone` to 1 if we entered. `curr.is_pad_zone === prev.is_pad_zone + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv)`
+5. For the 135th row
+   1. Same as the previous 0x80 pad, but pad 0x01 if we are in the pad zone. `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80 + is_pad_zone * 0x01`
+6. Use another RLC gadget to check `byte` can be running summed up to `input` in the lookup region
+7. `padded_byte` are copied to a word builder gadget to build padded words, which would later be copied to the `Keccak-f` permutation
```

Comment:
This is a bit confusing. As some of the big points overlap the rows they touch. We could maybe group them.
Also, there's no mention to the special padding case with `0x81`

Reply:
For big points 1~5, those checks have to be applied on their specific offsets. 6 is another gadget and 7 is all copy constraints.
I just realized point 4 can be grouped. Let me check.
The `0x81` case is in the 5.1. If it is a pad zone and condition_80 is activated then it would be `0x81`.

Reply:
Makes sense! Thanks!!!

Reply:
Addressed in [40e0186](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/40e0186951315ab16eca26dfcb932f0b15b19933)